docs/extra-chapter/hardware.md

```diff
@@ -0,0 +1,100 @@
+
+## Out of order execution
+All of this cache behavior brings a big amount of randomness into how long an instruction takes to load. The same `mov` instruction could take 10 cycles if it was on L1, but it could also take 200 if it was on RAM. During that time, the CPU would need to completely stop executing until that memory arrives. This happened on the PS3 cpu and before, and was one of its main sources of slowdown even while that cpu had a fast clock speed.
+To deal with this, CPUs began to execute instructions out of order. This way the pipeline of the cpu can continue doing useful work even if one of the instructions blocks for some time. Once out of order execution starts being a thing, the CPU can also improve performance by having multiple execution units for the different stages, and extract parallelism by itself. Modern CPUs can enqueue hundreds of instructions and schedule them for execution.
+
+For example, one single CPU core having 3 separate ALUs. One ALU is running a long operation like a divide, while the other ALUs are doing quick 1-cycle adds. But when doing that we run into a problem, which is that some instructions end at different times, and some instructions also depend on each other. This is handled internally in the CPU through some queing systems, where instructions get read multiple at a time, and then after they are decoded they get into a pending queue until their dependencies are met. Due to the low amount of registers used, the registers would constantly conflict with each other, so hardware designers added a "register renamer" to the CPU that connects the low amount of registers in the code into a much bigger amount of registers in the cpu to improve the way the instructions can overlap. Once a instruction has met its dependencies, it gets added into the different execution units of the cpu, such as a ALU that does adds/sub only, or an ALU that does divides too, or one of multiple memory loaders. 
+
+While instructions run at different amount of clock cycles and they also finish out of order, the cpu keeps track of an output list which will perform things like memory writes and atomic operations fully preserving the order of the code so that it works as written.
+
+Some info can be seen on this wikipedia page [4]
```

Comment:
```suggestion
All of this cache behavior brings a big amount of randomness into how long an instruction takes to load. The same `mov` instruction could take 10 cycles if it was in L1, but could take 200 cycles if it has to access RAM. During that time, the CPU would need to completely stop executing that instruction until the memory arrives. This happened to the PS3 CPU, and was one of its main sources of slowdowns, even though that CPU had a fast clock speed.
To deal with this, CPUs began to execute instructions out of order. This way the pipeline of the CPU can continue doing useful work even if one of the instructions has to wait for some time. Once out of order execution starts being a thing, the CPU can also improve performance by having multiple execution units for the different stages, and extract parallelism by itself. Modern CPUs can enqueue hundreds of instructions and schedule them for execution.

For example, one single CPU core could have 3 separate ALUs. One ALU is running a long operation like a divide, while the other ALUs are doing quick 1-cycle adds. But when doing that we run into a problem, which is that some instructions end at different times, and some instructions also depend on each other. This is handled internally in the CPU through queuing systems, where multiple instructions are read at a time, then after they are decoded they get put into a pending queue until their dependencies are met. Due to instructions using a low amount of registers, the registers required by different queued instructions would constantly conflict with each other, so hardware designers added a "register renamer" to the CPU that connects the low amount of registers used in the code into a much bigger amount of registers that the CPU has to improve instructions overlap. Once a instruction has met its dependencies, it gets added into the different execution units of the CPU, such as a ALU that does adds/sub only, or an ALU that does divides too, or one of multiple memory loaders. 

While instructions run at different amount of clock cycles and they also finish out of order, the CPU keeps track of an output list which will perform things like memory writes and atomic operations fully preserving the order of the code so that it works as written.

Some info can be seen on this wikipedia page [4]
```